ascent/examples/sequences_using_recursive_records.rs

```diff
@@ -0,0 +1,79 @@
+//! Sequences using Recursive Records
+//!
+//! The following example demonstrates the utilization of recursive records for
+//! building sequences of strings over a given alphabet.
+//!
+//! The Seq type is a recursive type where each instance is either nil or a pair of
+//! a leading `Letter` and a tailing `Seq`.
+//!
+//! The relation `seq` is defined to contain all sequences of length 5 or less
+//! over a given alphabet defined by the relation letter.
+//! The relation `len` is essentially a function assigning each sequence its length.
+//!
+//! Finally, the res relation illustrates how to create constant values for recursive record types.
+//!
+//! (Adapted from https://souffle-lang.github.io/examples#context-sensitive-flow-graph-with-records)
+
+use std::rc::Rc;
+
+use ascent::ascent;
+
+#[derive(Clone, Eq, PartialEq, Hash, Debug)]
+pub struct Letter(&'static str);
+
+#[derive(Clone, Eq, PartialEq, Hash, Debug)]
+pub struct Seq(Letter, Option<Rc<Seq>>);
```

Comment:
If we want to keep this example, I'd say this should be redefined as 
```
pub enum Seq {
  Cons(Letter, Rc<Seq>),
  Nil
}
```

Reply:
Good point! I've gone one further and …

1. Renamed `Seq` to `List` (to match the [LISP name](https://lisp-lang.org/learn/lists) of what it's effectively mimicking) and make it generic over the item (i.e. `List<T>`):

```rust
#[derive(Clone, Eq, PartialEq, Hash, Debug)]
pub enum List<T> {
  Cons(T, Rc<List<T>>),
  Nil
}
```

2. Removed `Letter(&'static str)` in favor of just `char` (and `relation letter(Letter` in favor of `relation char(char)`).

3. And to then show-case sub-clause macros (which so far were missing an example) I've added:

```rust
macro_rules! cons {
    ($h: expr, $t: expr) => {
        Rc::new(List::Cons($h, $t))
    }
}

macro_rules! nil {
    () => {
        Rc::new(List::Nil)
    }
}
```

Allowing for shrinking …

```rust
res(Res("abc")) <-- seq(Some(Rc::new(Seq(Letter("a"), Some(Rc::new(Seq(Letter("b"), Some(Rc::new(Seq(Letter("c"), None))))))))));
```

… to the much more readable:

```rust
res(Res("abc")) <-- list(cons!('a', cons!('b', cons!('c', nil!()))));
```

Reply:
Nice!